### ISSUE TITLE
What is the recommended way of using webdataset with pytorch-lightning and ddp?

### ISSUE THREAD

**Issue by isty2e** [tmbdev/webdataset#25]
I am trying to build a `LightningDataModule` using webdataset, however I have encountered some difficulties implementing, since previously I have never used `IterableDataset`.

The first thing I tried is [this example](https://github.com/tmbdev/webdataset-examples/tree/master/imagenet-lightning) in webdataset-examples: In `LightningDataModule`, I defined a `setup()` method which defines train/val/test datasets with `wds.Dataset`, and in `train/val/test_dataloader()` `DataLoader` instances are initialized with those datasets. This led to duplicate batching.

Then after googling a bit, I found [this comment](https://github.com/pytorch/pytorch/issues/38419#issuecomment-629387318) and [another comment](https://github.com/tmbdev/webdataset/issues/12#issuecomment-692278795), so apparently either letting the dataset itself do the batching, or using `wds.MultiDataset`. Just replacing `DataLoader` in `train/val/test_dataloader()` with `MultiDataset` resulted in the same batch duplication, so I suppose something is wrong. What is the recommended method for ddp in general?

**Comment by harpone** [comment]
Are you using the `seed_everyhting` function (recommended in PyTorch Lightning) and `ddp` parallel training?

That might set the same seed on each process => exact same shuffling... I had a nasty bug with that here: https://github.com/PyTorchLightning/pytorch-lightning/issues/3348

**Comment by isty2e** [comment]
> Are you using the `seed_everyhting` function (recommended in PyTorch Lightning) and `ddp` parallel training?
Just ddp, not `seed_everything`. For the time being, reproducibility is not a concern for me.
 
> That might set the same seed on each process => exact same shuffling... I had a nasty bug with that here: [PyTorchLightning/pytorch-lightning#3348](https://github.com/PyTorchLightning/pytorch-lightning/issues/3348)
To be precise, I want batches to be evenly distributed without duplication, like when a normal `Dataset` instead of `IterableDataset` is used.

**Comment by harpone** [comment]
Actually if you do multi-node ddp, then you need to use seed_everything to ensure the distributed models are initialized to be identical, unless there's a very recent update that fixes that (IMO it's pretty dangerous).

**Comment by tmbdev** [comment]
Some comments:

- WebDataset doesn't randomly distribute samples across nodes, but shards. For large training jobs and many nodes, you should split your dataset into a few hundred shards at least.
- For IterableDatasets, you should do batching inside the Dataset so that whole batches are transferred to the DataLoader; this isn't specific to WebDataset. However, if you want to mix data more, you can unbatch, shuffle, and rebatch in the training job (MultiDataset makes that particularly simple, but you can also do it with DataLoader)
- Which shards are selected is handled by three hooks in WebDataset:
  - the `reseed_hook` is called allowing you to reseed the random number generator (default: noop)
  - the `node_selection` hook is called on a list of shards, returning the node-specific subset (default: noop)
  - the `shard_selection` hook is called on that list of shards, returning the final subset of shards
  - the `shard_shuffle` hook is called to shuffle those shards

By default, training jobs on each node will see the entire dataset for training; that's because distributed frameworks have difficulties with different numbers of training batches from different nodes. If you don't want that behavior, you need to update the `node_selection` hook to select a subset of shards for each node.

This setup attempts to satisfy a number of different existing constraints as well as possible: the expectations of distributed frameworks, the constraint that distributed sequential I/O cannot guarantee exactly equal distribution of samples, and the existing APIs.

(For precise epochs and equal number of batches per worker, you need additional network communications somewhere; Tensorcom with PUBSUB sockets are a cleaner and more efficient solution.)

TODO:
- update the Lightning example to the latest API
- update the documentation to clarify

**Comment by tchaton** [comment]
Hey @tmbdev and al,

Awesome work with webdataset ! I was wondering where was your example with Pytorch Lightning.

Best regards,
Thomas Chaton.

**Comment by tchaton** [comment]
Hey @tmbdev @harpone,

I have seen there is current work adding webdataset to Pytorch. https://github.com/pytorch/pytorch/issues/38419

I was wondering if you would like to collaborate with Pytorch Lightning to provide better support to all users ?

Best,
T.C

**Comment by harpone** [comment]
@tchaton sure, although I really can't say how much time I can spend on that...

FYI everything is working fine for me with Pytorch Lightning + webdataset, after having some issues with identical minibatches between processes, dataset length and so on. This is on GPUs. Still having some issues on TPUs (with torch-xla), probably because torch-xla has their own parallel dataloader implementation.

Haven't really kept abreast of the pytorch version, but it seems somewhat different to @tmbdev 's version? Does someone have info of their current plans? The github issues are not very clear on the long term plan.

**Comment by tmbdev** [comment]
PyTorch is engaging in a substantial redesign of the entire I/O pipeline. That's a good thing to do, since there are many other limitations in the current design and APIs that a third party library like WebDataset can't address.

The new PyTorch design is functionally a superset of WebDataset and will support TarIterator and related functionality. Depending on what the PyTorch team decides to include, the WebDataset class might just be a convenience function in core PyTorch. If the PyTorch team decides not to include all the functionality of WebDataset in the core, then the WebDataset library will remain a small add-on library but reuse as much as possible from PyTorch.

Either way, the WebDataset format (POSIX tar files for datasets) will be supported by PyTorch, probably both natively and through the WebDataset API.

My guess is that the PyTorch API rewrite will take a while, and meanwhile, you can use WebDataset; except for naming, WebDataset is likely going to be a strict subset of PyTorch's future APIs.

WebDataset was working fine with PyTorch Lightning in the past; I'll have to rewrite and republish the examples that demonstrate how to use it. The trickiest part is just getting multi-node right, and WebDataset makes that pretty simple: you define a shard-splitting function that depends on the node id.

**Comment by hav4ik** [comment]
@harpone @isty2e Were you able to get a working solution for PyTorch XLA + TPU + Lightning?

**Comment by harpone** [comment]
Well I kinda gave up on torch-xla... things were just mysteriously slow and it's pretty hard to debug all the issues. So using CUDA only now.

**Comment by hav4ik** [comment]
@harpone I was able to adapt WebDataset for PyTorch-Lightning + TPU. However, I'm having mysterious runtime errors (that doesn't related to webdataset in any way), so I'm also giving up for now. When I have time, I will try to make a minimal working example.

**Comment by tmbdev** [comment]
Thanks for the feedback. I've put together some other examples (tmbdev/wds-distributed), and have also been working on OCRopus4 (an OCR engine that, among other things, uses webdataset for its data management). I'm getting back to trying to update this example now.
